File: app/src/main/java/com/lionsquare/comunidadkenna/activitys/LocationPickerActivity.java
package com.lionsquare.comunidadkenna.activitys;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.databinding.DataBindingUtil;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresApi;
import android.support.design.widget.Snackbar;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;
import com.google.android.gms.common.GooglePlayServicesNotAvailableException;
import com.google.android.gms.common.GooglePlayServicesRepairableException;
import com.google.android.gms.location.places.Place;
import com.google.android.gms.location.places.ui.PlacePicker;
import com.google.android.gms.maps.model.LatLng;
import com.google.firebase.iid.FirebaseInstanceId;
import com.lionsquare.comunidadkenna.R;
import com.lionsquare.comunidadkenna.api.ServiceApi;
import com.lionsquare.comunidadkenna.databinding.ActivityLocationPickerBinding;
import com.lionsquare.comunidadkenna.db.DbManager;
import com.lionsquare.comunidadkenna.model.CheckoutLogin;
import com.lionsquare.comunidadkenna.model.RecoverProfile;
import com.lionsquare.comunidadkenna.model.Register;
import com.lionsquare.comunidadkenna.model.User;
import com.lionsquare.comunidadkenna.utils.DialogGobal;
import com.lionsquare.comunidadkenna.utils.Preferences;
import com.lionsquare.comunidadkenna.utils.StatusBarUtil;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LocationPickerActivity extends AppCompatActivity implements View.OnClickListener, Callback<CheckoutLogin> {
    ActivityLocationPickerBinding binding;
    private static final int PLACE_PICKER_REQUEST = 1;
    private static final int PERMIS_LOCATION = 1;
    private Preferences preferences;
    private DbManager dbManager;

    private DialogGobal dialogGobal;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_location_picker);
        StatusBarUtil.darkMode(this);
        dialogGobal = new DialogGobal(this);
        initSetUp();

    }

    private void initSetUp() {
        preferences = new Preferences(this);
        dbManager = new DbManager(this).open();

        if (dbManager.getUser() != null) {
            Intent iMenu = new Intent(this, MenuActivity.class);
            startActivity(iMenu);
            finish();
        } else {
            binding.placeSearchDialogOkTV.setOnClickListener(this);
            binding.alpBtnStar.setVisibility(View.GONE);
            if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                verifyPermission();
            } else {
                binding.placeSearchDialogOkTV.setEnabled(true);
                checkoutLogin();
            }
        }


    }

    void checkoutLogin() {
        dialogGobal.progressIndeterminateStyle();
        ServiceApi serviceApi = ServiceApi.retrofit.create(ServiceApi.class);
        Call<CheckoutLogin> call = serviceApi.checkoutEmail(preferences.getEmail());
        call.enqueue(this);
    }

    @Override
    protected void onRestart() {
        super.onRestart();
        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            verifyPermission();
        } else {
            binding.placeSearchDialogOkTV.setEnabled(true);
        }

    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == PLACE_PICKER_REQUEST) {
            if (resultCode == RESULT_OK) {
                Place place = PlacePicker.getPlace(this, data);
                if (place != null) {
                    LatLng latLng = place.getLatLng();
                    sendPrefile(latLng);
                    binding.alpBtnStar.setVisibility(View.VISIBLE);
                    binding.placeSearchDialogOkTV.setVisibility(View.GONE);
                    binding.placeSearchDialogCancelTV.setVisibility(View.GONE);
                } else {
                    Log.e("error", "sdfsgrfger");

                }
            }
        }

    }

    private void locationPlacesIntent() {
        try {

            PlacePicker.IntentBuilder builder = new PlacePicker.IntentBuilder();
            startActivityForResult(builder.build(this), PLACE_PICKER_REQUEST);

        } catch (GooglePlayServicesRepairableException | GooglePlayServicesNotAvailableException e) {
            e.printStackTrace();
            Log.e("error lat", String.valueOf(e));
        }
    }

    @Override
    public void onClick(View v) {
        locationPlacesIntent();

    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    private void verifyPermission() {
        int writePermission = checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION);
        if (writePermission != PackageManager.PERMISSION_GRANTED) {
            requestPermission();
            binding.placeSearchDialogOkTV.setEnabled(false);
        } else {
            //saveComments();
            binding.placeSearchDialogOkTV.setEnabled(true);
            checkoutLogin();
        }
    }


    @RequiresApi(api = Build.VERSION_CODES.M)
    private void requestPermission() {
        if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.ACCESS_FINE_LOCATION)) {
            showSnackBar();
        } else {
            requestPermissions(new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, PERMIS_LOCATION);
        }
    }

    private void showSnackBar() {
        Snackbar snackbar = Snackbar.make(binding.alpRoot, R.string.permission_location, Snackbar.LENGTH_INDEFINITE);
        View snackBarView = snackbar.getView();
        snackBarView.setBackgroundColor(getResources().getColor(R.color.primaryColor));
        snackbar.setAction("Configurar", new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent();
                intent.setAction(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                Uri uri = Uri.fromParts("package", "com.lionsquare.kenna", null);
                intent.setData(uri);
                startActivity(intent);
            }
        }).show();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == PERMIS_LOCATION) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                //saveComments();
                binding.placeSearchDialogOkTV.setEnabled(true);
                checkoutLogin();
            } else {
                binding.placeSearchDialogOkTV.setEnabled(false);
                showSnackBar();
            }
        }
    }

    private void sendPrefile(final LatLng latLng) {
        dialogGobal.progressIndeterminateStyle();
        ServiceApi serviceApi = ServiceApi.retrofit.create(ServiceApi.class);
        Call<Register> call = serviceApi.registerProfile(preferences.getName(), preferences.getEmail()
                , preferences.getImagePerfil(), FirebaseInstanceId.getInstance().getToken(), preferences.getTypeLogin(), latLng.latitude, latLng.longitude);

        call.enqueue(new Callback<Register>() {
            @Override
            public void onResponse(Call<Register> call, Response<Register> response) {
                dialogGobal.dimmis();
                dbManager.insertUser(preferences.getName(), preferences.getEmail(), preferences.getImagePerfil(),
                        preferences.getCover(), preferences.getTypeLogin(), preferences.getTokenSosial(), "token Fire",
                        latLng.latitude, latLng.longitude);
                Intent iMenu = new Intent(LocationPickerActivity.this, MenuActivity.class);
                startActivity(iMenu);
                finish();
            }

            @Override
            public void onFailure(Call<Register> call, Throwable t) {
                dialogGobal.dimmis();
                Log.e("error de conexion", String.valueOf(t));

                new MaterialDialog.Builder(LocationPickerActivity.this)
                        .title(R.string.error)
                        .content(R.string.ocurrio_un_error_al_contectar)
                        .cancelable(false)
                        .negativeText(R.string.intentar_otra_vez)
                        .onNegative(new MaterialDialog.SingleButtonCallback() {
                            @Override
                            public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                                sendPrefile(latLng);
                            }
                        })
                        .positiveText(R.string.salir)
                        .onPositive(new MaterialDialog.SingleButtonCallback() {
                            @Override
                            public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                                finish();
                            }
                        })
                        .progressIndeterminateStyle(true)
                        .show();
            }
        });

    }

    @Override
    public void onResponse(Call<CheckoutLogin> call, Response<CheckoutLogin> response) {
        dialogGobal.dimmis();
        if (response.body().getSuccess() == 1) {
            // TODO: 31/07/2017 actulizamos el perfil ya sea que cambio de cuanta
            if (preferences.getTypeLogin() != response.body().getType_account())
                diferenteAccount();
            else
                recoverProfileData();
        }
    }

    @Override
    public void onFailure(Call<CheckoutLogin> call, Throwable t) {
        Log.e("error de conexion", String.valueOf(t));
        dialogGobal.dimmis();
    }

    // TODO: 31/07/2017 si se accede con la misma cuenta solo se recuperan los datos del servidor con el correo
    void recoverProfileData() {
        dialogGobal.setDialog(getResources().getString(R.string.recuperando_datos));
        ServiceApi serviceApi = ServiceApi.retrofit.create(ServiceApi.class);
        Call<RecoverProfile> call = serviceApi.recoverProfile(preferences.getEmail());
        call.enqueue(new Callback<RecoverProfile>() {
            @Override
            public void onResponse(Call<RecoverProfile> call, Response<RecoverProfile> response) {
                dialogGobal.dimmis();
                if (response.body().getSuccess() == 1) {
                    User user = response.body().getUser();
                    dbManager.insertUser(preferences.getName(), preferences.getEmail(), preferences.getImagePerfil(),
                            preferences.getCover(), preferences.getTypeLogin(), preferences.getTokenSosial(), user.getToken(),
                            user.getLat(), user.getLat());
                    Intent iMenu = new Intent(LocationPickerActivity.this, MenuActivity.class);
                    startActivity(iMenu);
                    finish();
                    dbManager.close();
                }
            }

            @Override
            public void onFailure(Call<RecoverProfile> call, Throwable t) {
                dialogGobal.dimmis();
                Log.e("error", String.valueOf(t));
                dialogGobal.errorConexionFinish(LocationPickerActivity.this);
            }
        });

    }


    // TODO: 31/07/2017 cuando es iferentes ala cuenta con la que estabas pero es le mismo coreo se aztulizan los perfiles
    void diferenteAccount() {
        ServiceApi serviceApi = ServiceApi.retrofit.create(ServiceApi.class);
        Call<RecoverProfile> call = serviceApi.updateProfile(
                preferences.getEmail(), preferences.getName(), preferences.getImagePerfil(), FirebaseInstanceId.getInstance().getToken(), preferences.getTypeLogin());
        call.enqueue(new Callback<RecoverProfile>() {
            @Override
            public void onResponse(Call<RecoverProfile> call, Response<RecoverProfile> response) {
                if (response.body().getSuccess() == 1) {
                    User user = response.body().getUser();
                    dbManager.insertUser(preferences.getName(), preferences.getEmail(), preferences.getImagePerfil(),
                            preferences.getCover(), preferences.getTypeLogin(), preferences.getTokenSosial(), user.getToken(),
                            user.getLat(), user.getLat());
                    Intent iMenu = new Intent(LocationPickerActivity.this, MenuActivity.class);
                    startActivity(iMenu);
                    finish();
                    dbManager.close();
                }
            }

            @Override
            public void onFailure(Call<RecoverProfile> call, Throwable t) {
                dialogGobal.errorConexionFinish(LocationPickerActivity.this);
            }
        });
    }
}

File: app/src/main/java/com/lionsquare/comunidadkenna/activitys/LocationPickerActivity.kt
package com.lionsquare.comunidadkenna.activitys

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.databinding.DataBindingUtil
import android.net.Uri
import android.os.Build
import android.provider.Settings
import android.support.annotation.RequiresApi
import android.support.design.widget.Snackbar
import android.support.v4.app.ActivityCompat
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View

import com.afollestad.materialdialogs.DialogAction
import com.afollestad.materialdialogs.MaterialDialog
import com.google.android.gms.common.GooglePlayServicesNotAvailableException
import com.google.android.gms.common.GooglePlayServicesRepairableException
import com.google.android.gms.location.places.Place
import com.google.android.gms.location.places.ui.PlacePicker
import com.google.android.gms.maps.model.LatLng
import com.google.firebase.iid.FirebaseInstanceId
import com.lionsquare.comunidadkenna.R
import com.lionsquare.comunidadkenna.api.ServiceApi
import com.lionsquare.comunidadkenna.databinding.ActivityLocationPickerBinding
import com.lionsquare.comunidadkenna.db.DbManager
import com.lionsquare.comunidadkenna.model.CheckoutLogin
import com.lionsquare.comunidadkenna.model.RecoverProfile
import com.lionsquare.comunidadkenna.model.Register
import com.lionsquare.comunidadkenna.model.User
import com.lionsquare.comunidadkenna.utils.DialogGobal
import com.lionsquare.comunidadkenna.utils.Preferences
import com.lionsquare.comunidadkenna.utils.StatusBarUtil

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import thebat.lib.validutil.ValidUtils

class LocationPickerActivity : AppCompatActivity(), View.OnClickListener, Callback<CheckoutLogin> {
    internal lateinit var binding: ActivityLocationPickerBinding
    private var preferences: Preferences? = null
    private var dbManager: DbManager? = null

    private var dialogGobal: DialogGobal? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_location_picker)
        StatusBarUtil.darkMode(this)
        dialogGobal = DialogGobal(this)
        initSetUp()

    }

    private fun initSetUp() {
        preferences = Preferences(this)
        dbManager = DbManager(this).open()
        //  Log.e("token",FirebaseInstanceId.getInstance().getToken());

        if (dbManager!!.user != null) {
            val iMenu = Intent(this, MenuActivity::class.java)
            startActivity(iMenu)
            finish()
        } else {
            binding.placeSearchDialogOkTV.setOnClickListener(this)

            if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                verifyPermission()
            } else {
                binding.placeSearchDialogOkTV.isEnabled = true
                if (ValidUtils.isNetworkAvailable(this))
                    checkoutLogin()
                else {

                    dialogGobal!!.sinInternet(this)

                }

            }
        }

        binding.placeSearchDialogCancelTV.setOnClickListener(this)


    }

    internal fun checkoutLogin() {
        dialogGobal!!.progressIndeterminateStyle()
        val serviceApi = ServiceApi.retrofit.create(ServiceApi::class.java)
        val call = serviceApi.checkoutEmail(preferences!!.email)
        call.enqueue(this)
    }

    override fun onRestart() {
        super.onRestart()
        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            verifyPermission()
        } else {
            binding.placeSearchDialogOkTV.isEnabled = true
        }

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        if (requestCode == PLACE_PICKER_REQUEST) {
            if (resultCode == Activity.RESULT_OK) {
                val place = PlacePicker.getPlace(this, data)
                if (place != null) {
                    val latLng = place.latLng
                    if (ValidUtils.isNetworkAvailable(this))
                        sendPrefile(latLng)
                    else
                        dialogGobal!!.sinInternet(this)

                    binding.placeSearchDialogOkTV.visibility = View.GONE
                    binding.placeSearchDialogCancelTV.visibility = View.GONE
                } else {
                    Log.e("error", "sdfsgrfger")

                }
            }
        }

    }

    private fun locationPlacesIntent() {
        try {

            val builder = PlacePicker.IntentBuilder()

            startActivityForResult(builder.build(this), PLACE_PICKER_REQUEST)

        } catch (e: GooglePlayServicesRepairableException) {
            e.printStackTrace()
            Log.e("error lat", e.toString())
        } catch (e: GooglePlayServicesNotAvailableException) {
            e.printStackTrace()
            Log.e("error lat", e.toString())
        }

    }

    override fun onClick(v: View) {

        when (v.id) {
            R.id.place_search_dialog_cancel_TV -> finish()
            R.id.place_search_dialog_ok_TV -> locationPlacesIntent()
        }

    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    private fun verifyPermission() {
        val writePermission = checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION)
        if (writePermission != PackageManager.PERMISSION_GRANTED) {
            requestPermission()
            binding.placeSearchDialogOkTV.isEnabled = false
        } else {
            //saveComments();
            binding.placeSearchDialogOkTV.isEnabled = true
            checkoutLogin()
        }
    }


    @RequiresApi(api = Build.VERSION_CODES.M)
    private fun requestPermission() {
        if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.ACCESS_FINE_LOCATION)) {
            showSnackBar()
        } else {
            requestPermissions(arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), PERMIS_LOCATION)
        }
    }

    private fun showSnackBar() {
        val snackbar = Snackbar.make(binding.alpRoot, R.string.permission_location, Snackbar.LENGTH_INDEFINITE)
        val snackBarView = snackbar.view
        snackBarView.setBackgroundColor(resources.getColor(R.color.primaryColor))
        snackbar.setAction("Configurar") {
            val intent = Intent()
            intent.action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS
            val uri = Uri.fromParts("package", "com.lionsquare.kenna", null)
            intent.data = uri
            startActivity(intent)
        }.show()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == PERMIS_LOCATION) {
            if (grantResults.size > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                //saveComments();
                binding.placeSearchDialogOkTV.isEnabled = true
                checkoutLogin()
            } else {
                binding.placeSearchDialogOkTV.isEnabled = false
                showSnackBar()
            }
        }
    }

    private fun sendPrefile(latLng: LatLng) {
        dialogGobal!!.progressIndeterminateStyle()
        val serviceApi = ServiceApi.retrofit.create(ServiceApi::class.java)
        val call = serviceApi.registerProfile(preferences!!.name, preferences!!.email, preferences!!.imagePerfil, FirebaseInstanceId.getInstance().token, preferences!!.typeLogin, latLng.latitude, latLng.longitude)

        call.enqueue(object : Callback<Register> {
            override fun onResponse(call: Call<Register>, response: Response<Register>) {
                dialogGobal!!.dimmis()
                dbManager!!.insertUser(preferences!!.name, preferences!!.email, preferences!!.imagePerfil,
                        preferences!!.cover, preferences!!.typeLogin, preferences!!.tokenSosial, preferences!!.token,
                        latLng.latitude, latLng.longitude)
                val iMenu = Intent(this@LocationPickerActivity, MenuActivity::class.java)
                startActivity(iMenu)
                finish()
            }

            override fun onFailure(call: Call<Register>, t: Throwable) {
                dialogGobal!!.dimmis()
                Log.e("error de conexion", t.toString())

                MaterialDialog.Builder(this@LocationPickerActivity)
                        .title(R.string.error)
                        .content(R.string.ocurrio_un_error_al_contectar)
                        .cancelable(false)
                        .negativeText(R.string.intentar_otra_vez)
                        .onNegative { dialog, which ->
                            if (!ValidUtils.isNetworkAvailable(this@LocationPickerActivity))
                                sendPrefile(latLng)
                            else
                                dialogGobal!!.sinInternet(this@LocationPickerActivity)
                        }
                        .positiveText(R.string.salir)
                        .onPositive { dialog, which -> finish() }
                        .progressIndeterminateStyle(true)
                        .show()
            }
        })

    }

    override fun onResponse(call: Call<CheckoutLogin>, response: Response<CheckoutLogin>) {
        dialogGobal!!.dimmis()
        if (response.body().success == 1) {
            // TODO: 31/07/2017 actulizamos el perfil ya sea que cambio de cuanta
            if (preferences!!.typeLogin != response.body().type_account)
                diferenteAccount()
            else
                recoverProfileData()
        }
    }

    override fun onFailure(call: Call<CheckoutLogin>, t: Throwable) {
        Log.e("error de conexion", t.toString())
        dialogGobal!!.dimmis()
    }

    // TODO: 31/07/2017 si se accede con la misma cuenta solo se recuperan los datos del servidor con el correo
    internal fun recoverProfileData() {
        dialogGobal!!.setDialog(resources.getString(R.string.recuperando_datos))
        val serviceApi = ServiceApi.retrofit.create(ServiceApi::class.java)
        val call = serviceApi.recoverProfile(preferences!!.email, FirebaseInstanceId.getInstance().token)
        call.enqueue(object : Callback<RecoverProfile> {
            override fun onResponse(call: Call<RecoverProfile>, response: Response<RecoverProfile>) {
                dialogGobal!!.dimmis()
                if (response.body().success == 1) {
                    val user = response.body().user
                    dbManager!!.insertUser(preferences!!.name, preferences!!.email, preferences!!.imagePerfil,
                            preferences!!.cover, preferences!!.typeLogin, preferences!!.tokenSosial, user.token,
                            user.lat, user.lng)
                    val iMenu = Intent(this@LocationPickerActivity, MenuActivity::class.java)
                    startActivity(iMenu)
                    finish()
                    dbManager!!.close()
                }
            }

            override fun onFailure(call: Call<RecoverProfile>, t: Throwable) {
                dialogGobal!!.dimmis()
                Log.e("error", t.toString())
                dialogGobal!!.errorConexionFinish(this@LocationPickerActivity)
            }
        })

    }


    // TODO: 31/07/2017 cuando es iferentes ala cuenta con la que estabas pero es le mismo coreo se aztulizan los perfiles
    internal fun diferenteAccount() {
        val serviceApi = ServiceApi.retrofit.create(ServiceApi::class.java)
        val call = serviceApi.updateProfile(
                preferences!!.email, preferences!!.name, preferences!!.imagePerfil, FirebaseInstanceId.getInstance().token, preferences!!.typeLogin)
        call.enqueue(object : Callback<RecoverProfile> {
            override fun onResponse(call: Call<RecoverProfile>, response: Response<RecoverProfile>) {
                if (response.body().success == 1) {
                    val user = response.body().user
                    dbManager!!.insertUser(preferences!!.name, preferences!!.email, preferences!!.imagePerfil,
                            preferences!!.cover, preferences!!.typeLogin, preferences!!.tokenSosial, user.token,
                            user.lat, user.lng)
                    val iMenu = Intent(this@LocationPickerActivity, MenuActivity::class.java)
                    startActivity(iMenu)
                    finish()
                    dbManager!!.close()
                }
            }

            override fun onFailure(call: Call<RecoverProfile>, t: Throwable) {
                dialogGobal!!.errorConexionFinish(this@LocationPickerActivity)
            }
        })
    }

    companion object {
        private val PLACE_PICKER_REQUEST = 1
        private val PERMIS_LOCATION = 1
    }
}
